src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-                return emittedWhileDraining;
-            }
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
+                    add(svq);
+                    scalarQueue = svq;
+                }
+                try {
+                    svq.onNext(nl.next(value));
+                } catch (MissingBackpressureException ex) {
+                    onError(ex);
+                    return;
+                } catch (IllegalStateException ex) {
+                    // if there was an IAE and we are unsubscribed, just quit
+                    if (!isUnsubscribed()) {
+                        onError(ex);
                     }
+                    return;
                 }
-                return Boolean.TRUE;
-            }
-
-        };
-
-        @Override
-        public void onError(Throwable e) {
-            if (!completed) {
-                completed = true;
-                innerError(e, true);
+                // then try to emit it (or notify the running emission loop).
+                emit();
             }
         }
         
-        private void innerError(Throwable e, boolean parent) {
-            if (delayErrors) {
-                synchronized (this) {
-                    if (exceptions == null) {
-                        exceptions = new ConcurrentLinkedQueue<Throwable>();
+        void fastEmitNext(T value) {
+            synchronized (this) {
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
```

Comment:
It is unbounded down, that doesn't mean unbounded up. The `merge` operator is an async operator, just like `zip` and `observeOn` since it does queueing. This means the `merge` operator itself IS an async operator and thus has buffered, `request(n)` semantics regardless of whether the downstream requests `Long.MAX_VALUE` or not.

The only cases I'm aware of where it doesn't need to buffer are the following where no concurrent emission occurs:

1) When `merge` has only a single stream to merge (not a real merge, so not an interesting case)
2) When `maxConcurrent==1` which is the same as `concat` or effectively the same as usecase 1 above.
3) When each Observable being merged is synchronous, which can't be known by `merge` ahead of time. In this case it again behaves like `concat`. 

Since (3) is the common case to optimize for (ie. many short synchronous streams), we would have to determine this at runtime. Basically we assume that if we never have `missed` occur due to concurrent emission, we stay on the "fast path", but as soon as we see concurrency occur we have to use queues.

This is why the current implementation has an optimization to skip the queues: https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/internal/operators/OperatorMerge.java#L612 and https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/internal/operators/OperatorMerge.java#L686

But it uses the queues as soon as concurrent emissions happen.